---
title: "Emma Pan Neuro (Control + ND75KD) - Find origin of Clusters 2/23"
author: "Vincent Gardeux"
date: "2024-08-12"
output:
  html_document:
    df_print: paged
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/updepla_storage/users/gardeux/2023-04-19_snRNAseq_Emma/analysis/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat)) # For single-cell pipeline
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(loomR)) # For building a file for ASAP
suppressPackageStartupMessages(library(Matrix)) # For building a file for ASAP
suppressPackageStartupMessages(library(ggplot2)) # For plotting
suppressPackageStartupMessages(library(ggpubr)) # For grid plotting
suppressPackageStartupMessages(library(plotly)) # For interactive plots
suppressPackageStartupMessages(library(RColorBrewer)) # For nice color Palettes
suppressPackageStartupMessages(library(harmony)) # For integration
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("loomR"), "version", as.character(packageVersion("loomR")), "\n")
cat(bold("Matrix"), "version", as.character(packageVersion("Matrix")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("ggpubr"), "version", as.character(packageVersion("ggpubr")), "\n")
cat(bold("plotly"), "version", as.character(packageVersion("plotly")), "\n")
cat(bold("harmony"), "version", as.character(packageVersion("harmony")), "\n")
cat(bold("RColorBrewer"), "version", as.character(packageVersion("RColorBrewer")), "\n")

# Parameters
seurat_integrated <- "Pan_neuro_both_reannotated_GFP_curated_reintegrated_SCENIC_binarized_FLYBASETF.rds"

# Color-blind friendly palette 3-colors
cbPalette <- c("#E69F00", "#000000", "#56B4E9")
steinPalette <- c("#80C980", "#BDADD4", "#376CB0", "#FBBF85", "#F0027E")

# Random seed
set.seed(42)
```

## I.1 Reading Seurat object

First step is to read the Seurat objects, previously created

```{r}
message("Loading Seurat object...")
data.seurat <- readRDS(seurat_integrated)
message(ncol(data.seurat), " cells were loaded")
message(nrow(data.seurat), " genes were loaded")
```

## I.2 A few basic plots

```{r, fig.height = 12, fig.width = 14}
p <- DimPlot(data.seurat, reduction = "tsne_harmony", label = F, pt.size = 0.5, group.by = "orig.ident", cols = cbPalette[c(1,3)])
p
```

```{r, fig.height = 12, fig.width = 14}
p <- DimPlot(data.seurat, reduction = "umap_harmony", label = F, pt.size = 0.5, group.by = "orig.ident", cols = cbPalette[c(1,3)])
p
```

```{r, fig.height = 12, fig.width = 13}
all_levels <- unique(data.seurat$annotation_final_merged)
default_colors <- colorRampPalette(brewer.pal(12, "Paired"))(length(all_levels))
names(default_colors) <- all_levels

p <- DimPlot(data.seurat, reduction = "tsne_harmony", label = F, pt.size = 0.5, group.by = "annotation_final_merged", cols = default_colors) + NoLegend()
p <- LabelClusters(p, id = "annotation_final_merged",  fontface = "bold", repel = F)
p
```

```{r, fig.height = 12, fig.width = 13}
all_levels <- unique(data.seurat$annotation_final_merged)
default_colors <- colorRampPalette(brewer.pal(12, "Paired"))(length(all_levels))
names(default_colors) <- all_levels

# Generate the DimPlot with these specified colors
p <- DimPlot(data.seurat, reduction = "umap_harmony", label = F, pt.size = 0.5, group.by = "annotation_final_merged", cols = default_colors) + NoLegend()
p <- LabelClusters(p, id = "annotation_final_merged",  fontface = "bold", repel = F)
p
```

## II.1. First: re-integrate and force integration of these clusters into the FCA data
```{r}
# Read FCA object
data.seurat_fca <- readRDS("Fly_Cell_Atlas__Head_Stringent.rds")
message(ncol(data.seurat_fca), " cells were loaded")
message(nrow(data.seurat_fca), " genes were loaded")

# Create a single Seurat object
data.seurat_reintegrated <- merge(x = data.seurat, y = data.seurat_fca, project = "Pan_neuro_FCA_reintegrated")

# Filling NAs in metadata
data.seurat_reintegrated$batch[data.seurat_reintegrated$orig.ident == "Pan_neuro_control"] <- 13
data.seurat_reintegrated$batch[data.seurat_reintegrated$orig.ident == "Pan_neuro_ND75KD"] <- 14

# Size of the new Seurat object
message(ncol(data.seurat_reintegrated), " cells in integrated Seurat object")
message(nrow(data.seurat_reintegrated), " genes in total (union)")

# By default, all non-overlapping genes are set to 0. Removing genes that are not overlapping between the two datasets.
data.seurat_reintegrated <- data.seurat_reintegrated[sort(intersect(rownames(data.seurat_fca), rownames(data.seurat))),]

# Size of the new Seurat object
message(nrow(data.seurat_reintegrated), " genes in integrated Seurat object (intersection)")

# Default pipeline
data.seurat_reintegrated <- NormalizeData(data.seurat_reintegrated, normalization.method = "LogNormalize", scale.factor = 10000, verbose = F)
data.seurat_reintegrated <- FindVariableFeatures(data.seurat_reintegrated, selection.method = "vst", nfeatures = 2000, verbose = F)
data.seurat_reintegrated <- ScaleData(data.seurat_reintegrated, verbose = F)
data.seurat_reintegrated <- RunPCA(data.seurat_reintegrated, features = VariableFeatures(data.seurat_reintegrated), verbose = F, npcs = 100)
```

```{r}
# Run Harmony
# Here I force Harmony to run on 30 iterations, without early stop. To try forcing the integration of the 2/23 cluster
data.seurat_reintegrated <- RunHarmony(object = data.seurat_reintegrated, group.by.vars = "batch", plot_convergence = TRUE, max_iter = 30, early_stop = F, verbose = F)
```
```{r}
# Clustering
data.seurat_reintegrated <- FindNeighbors(data.seurat_reintegrated, dims = 1:100, reduction = "harmony", verbose = F)
# I use a resolution of 4, which is completely arbitrary, we may tune it to have more/less clusters
data.seurat_reintegrated <- FindClusters(data.seurat_reintegrated, resolution = 4)

# UMAP (visualization)
data.seurat_reintegrated <- RunUMAP(data.seurat_reintegrated, dims = 1:100, reduction = "harmony", verbose = F)
DimPlot(data.seurat_reintegrated, reduction = "umap", pt.size = 1, group.by = "batch")
```

```{r, fig.height = 24, fig.width = 26}
# UMAP (visualization)
# Clusters 2/23 are now both in cluster 45
DimPlot(data.seurat_reintegrated, reduction = "umap", pt.size = 1, group.by = "seurat_clusters")
```

```{r, fig.height = 24, fig.width = 26}
data.seurat_reintegrated$annotation_final_merged[is.na(data.seurat_reintegrated$annotation_final_merged)] <- paste0("FCA_", data.seurat_reintegrated$annotation[is.na(data.seurat_reintegrated$annotation_final_merged)])
all_levels <- unique(data.seurat_reintegrated$annotation_final_merged)
default_colors <- colorRampPalette(brewer.pal(12, "Paired"))(length(all_levels))
names(default_colors) <- all_levels

# Generate the DimPlot with these specified colors
p <- DimPlot(data.seurat_reintegrated, reduction = "umap", label = F, pt.size = 0.5, group.by = "annotation_final_merged", cols = default_colors, raster = F) + NoLegend()
p <- LabelClusters(p, id = "annotation_final_merged",  fontface = "bold", repel = F)
p
```

It seems it did not merge again... we can see it at the top, near the Poxn / OPN neurons clusters

## II.2 Hierachical clustering

### II.2.1 On integrated Seurat clusters (FCA + Pan neuro)

```{r}
# 1. Summarize the data by cluster
clusters <- as.character(sort(unique(data.seurat_reintegrated$seurat_clusters)))
data.results <- data.frame(matrix(0, nrow = length(clusters), ncol = ncol(data.seurat_reintegrated@reductions$harmony@cell.embeddings)), row.names = clusters)
colnames(data.results) <- colnames(data.seurat_reintegrated@reductions$harmony@cell.embeddings)
for(clust in clusters){
  data.results[clust,] <- colMeans(data.seurat_reintegrated@reductions$harmony@cell.embeddings[data.seurat_reintegrated$seurat_clusters == clust,])
}
```

```{r, fig.height = 5, fig.width = 20}
# 2. Hierarchical clustering
hc <- hclust(dist(data.results, method = "euclidean"), method = "ward")
hclustplot(hc, k = 40)
```

Cluster 45 is close to the big blob in the top-center of the integrated UMAP (109, 71, 90 ,2 ,3)

### II.2.2 On Emma + FCA annotation (FCA + Pan neuro)

```{r}
# 1. Summarize the data by cluster
clusters <- as.character(sort(unique(data.seurat_reintegrated$annotation_final_merged)))
data.results <- data.frame(matrix(0, nrow = length(clusters), ncol = ncol(data.seurat_reintegrated@reductions$harmony@cell.embeddings)), row.names = clusters)
colnames(data.results) <- colnames(data.seurat_reintegrated@reductions$harmony@cell.embeddings)
for(clust in clusters){
  data.results[clust,] <- colMeans(data.seurat_reintegrated@reductions$harmony@cell.embeddings[data.seurat_reintegrated$annotation_final_merged == clust,])
}
```

```{r, fig.height = 8, fig.width = 20}
# 2. Hierarchical clustering
hc <- hclust(dist(data.results, method = "euclidean"), method = "ward")
hclustplot(hc, k = 40, mar = c(15,1,1,1))
```

# Different clustering method (mcquitty i.e. WPGMA)

```{r, fig.height = 8, fig.width = 20}
# 2. Hierarchical clustering
hc <- hclust(dist(data.results, method = "euclidean"), method = "mcquitty")
hclustplot(hc, k = 50, mar = c(15,1,1,1))
```

### II.2.3 On Pan neuro only (Emma annotations)

```{r}
# 1. Summarize the data by cluster
clusters <- as.character(sort(unique(data.seurat$annotation_final_merged)))
data.results <- data.frame(matrix(0, nrow = length(clusters), ncol = ncol(data.seurat@reductions$harmony@cell.embeddings)), row.names = clusters)
colnames(data.results) <- colnames(data.seurat@reductions$harmony@cell.embeddings)
for(clust in clusters){
  data.results[clust,] <- colMeans(data.seurat@reductions$harmony@cell.embeddings[data.seurat$annotation_final_merged == clust,])
}
```

```{r, fig.height = 6, fig.width = 12}
# 2. Hierarchical clustering
hc <- hclust(dist(data.results, method = "euclidean"), method = "ward")
hclustplot(hc, k = 40, mar = c(15,1,1,1))
```