{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "69fcd067",
   "metadata": {},
   "source": [
    "# Emma Pan Neuro (Control + ND75KD) - pySCENIC pipeline (CLI version)\n",
    "\n",
    "**Author:** Vincent Gardeux\n",
    "\n",
    "**Date Created:** 03/01/2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c66bc4a6",
   "metadata": {},
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b9c1fa0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import dependencies\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import loompy as lp\n",
    "from MulticoreTSNE import MulticoreTSNE as TSNE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03e94404",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "beaef2b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.10.3\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "MulticoreTSNE       NA\n",
      "PIL                 8.2.0\n",
      "argcomplete         NA\n",
      "asttokens           NA\n",
      "astunparse          1.6.3\n",
      "awkward             2.3.1\n",
      "awkward_cpp         NA\n",
      "backcall            0.2.0\n",
      "cairo               1.21.0\n",
      "cffi                1.14.5\n",
      "cloudpickle         2.2.0\n",
      "colorama            0.4.6\n",
      "comm                0.1.3\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "cytoolz             0.10.1\n",
      "dask                2021.06.0\n",
      "dateutil            2.8.2\n",
      "debugpy             1.6.7\n",
      "decorator           4.4.2\n",
      "dill                0.3.4\n",
      "entrypoints         0.3\n",
      "exceptiongroup      1.1.3\n",
      "executing           1.2.0\n",
      "fsspec              2023.4.0\n",
      "future_fstrings     NA\n",
      "google              NA\n",
      "h5py                3.7.0\n",
      "igraph              0.10.4\n",
      "ipykernel           6.25.0\n",
      "ipython_genutils    0.2.0\n",
      "ipywidgets          7.6.3\n",
      "isal                1.5.2\n",
      "jedi                0.18.0\n",
      "joblib              1.2.0\n",
      "kiwisolver          1.3.1\n",
      "leidenalg           0.9.1\n",
      "llvmlite            0.39.1\n",
      "loompy              3.0.7\n",
      "louvain             0.7.0\n",
      "lz4                 3.1.3\n",
      "matplotlib          3.5.1\n",
      "matplotlib_inline   NA\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "numba               0.56.3\n",
      "numexpr             2.7.3\n",
      "numpy               1.23.5\n",
      "numpy_groupies      0.9.13\n",
      "opt_einsum          v3.3.0\n",
      "packaging           23.1\n",
      "pandas              1.5.3\n",
      "parso               0.8.2\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "platformdirs        3.9.1\n",
      "plotly              5.13.1\n",
      "prompt_toolkit      3.0.39\n",
      "psutil              5.9.4\n",
      "ptyprocess          0.7.0\n",
      "pure_eval           0.2.2\n",
      "pyarrow             15.0.0\n",
      "pycparser           2.20\n",
      "pydev_ipython       NA\n",
      "pydevconsole        NA\n",
      "pydevd              2.9.5\n",
      "pydevd_file_utils   NA\n",
      "pydevd_plugins      NA\n",
      "pydevd_tracing      NA\n",
      "pyexpat             NA\n",
      "pygments            2.15.1\n",
      "pyparsing           2.4.7\n",
      "pytz                2021.1\n",
      "ruamel              NA\n",
      "scipy               1.11.1\n",
      "session_info        1.0.0\n",
      "setuptools_scm      NA\n",
      "six                 1.15.0\n",
      "sklearn             1.3.1\n",
      "sparse              0.14.0\n",
      "sphinxcontrib       NA\n",
      "stack_data          0.6.2\n",
      "tblib               1.7.0\n",
      "texttable           1.6.3\n",
      "threadpoolctl       2.1.0\n",
      "tlz                 0.10.1\n",
      "toolz               0.11.1\n",
      "torch               1.13.0+cu117\n",
      "tornado             6.3.1\n",
      "tqdm                4.65.0\n",
      "traitlets           5.9.0\n",
      "typing_extensions   NA\n",
      "wcwidth             0.2.5\n",
      "xxhash              2.0.2\n",
      "yaml                6.0.1\n",
      "zmq                 24.0.1\n",
      "zope                NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      7.4.9\n",
      "jupyter_core        5.3.1\n",
      "notebook            6.5.5\n",
      "-----\n",
      "Python 3.9.5 (default, Jun 15 2022, 08:55:22) [GCC 8.5.0 20210514 (Red Hat 8.5.0-10)]\n",
      "Linux-4.18.0-477.15.1.el8_8.x86_64-x86_64-with-glibc2.28\n",
      "-----\n",
      "Session information updated at 2024-03-06 13:55\n"
     ]
    }
   ],
   "source": [
    "# Parameters for scanpy\n",
    "sc.settings.verbosity = 3 # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "sc.set_figure_params(dpi=150, fontsize=10, dpi_save=600)\n",
    "sc.settings.njobs = 32 # Set maximum number of jobs for Scanpy (also used in GRNBoost2)\n",
    "\n",
    "# [Input] Loom file to use\n",
    "f_loom_path_scenic = \"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/analysis/Pan_neuro_both_reannotated_GFP_curated_reintegrated.loom\"\n",
    "\n",
    "# [Input] Transcription factors list (SCENIC step 1: GRNBoost2)\n",
    "f_tfs = \"/data/genome/drosophila_melanogaster/cistopic_flybase_r6.02/allTFs_dmel.txt\" # drosophila\n",
    "\n",
    "# [Output] Adjacency matrix (SCENIC step 1: GRNBoost2)\n",
    "adj_matrix = \"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/analysis/Pan_neuro_both_reannotated_GFP_curated_reintegrated_adj.csv\"\n",
    "\n",
    "# [Input] Ranking databases (SCENIC step 2-3: cisTarget)\n",
    "f_db_names = \"/data/genome/drosophila_melanogaster/cistopic_flybase_r6.02/mc_v10_clust/dm6_v10_clust.genes_vs_motifs.rankings.feather\"\n",
    "\n",
    "# [Input] Motif databases (SCENIC step 2-3: cisTarget)\n",
    "f_motif_path = \"/data/genome/drosophila_melanogaster/cistopic_flybase_r6.02/mc_v10_clust/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl\"\n",
    "\n",
    "# [Output] Regulons (SCENIC step 2-3: cisTarget)\n",
    "reg_matrix = \"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/analysis/Pan_neuro_both_reannotated_GFP_curated_reintegrated_reg.csv\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a56c225",
   "metadata": {},
   "source": [
    "# SCENIC steps\n",
    "\n",
    "## STEP 1: Gene regulatory network inference, and generation of co-expression modules\n",
    "\n",
    "### 1.a. GRN inference using the GRNBoost2 algorithm\n",
    "\n",
    "For this step the CLI version of SCENIC is used. This step can be deployed on an High Performance Computing system. We use the counts matrix (without log transformation or further processing) from the loom file we wrote earlier. \n",
    "\n",
    "*Output:* List of adjacencies between a TF and its targets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a89ed6",
   "metadata": {},
   "source": [
    "Run GRNBoost2 algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a355d9ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 11:00:06,775 - pyscenic.cli.pyscenic - INFO - Loading expression matrix.\n",
      "\n",
      "2024-03-06 11:00:23,383 - pyscenic.cli.pyscenic - INFO - Inferring regulatory networks.\n",
      "preparing dask client\n",
      "parsing input\n",
      "creating dask graph\n",
      "32 partitions\n",
      "computing dask graph\n",
      "not shutting down client, client was created externally\n",
      "finished\n",
      "\n",
      "2024-03-06 12:44:45,916 - pyscenic.cli.pyscenic - INFO - Writing results to file.\n"
     ]
    }
   ],
   "source": [
    "# Here I run the CLI version from the Docker\n",
    "!docker run -it --rm --name pySCENIC \\\n",
    "    -v /data/genome/drosophila_melanogaster/cistopic_flybase_r6.02:/cistopic_flybase_r6.02 \\\n",
    "    -v /home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/analysis:/analysis \\\n",
    "    -w /analysis aertslab/pyscenic:0.12.1 \\\n",
    "    pyscenic grn \\\n",
    "    {f_loom_path_scenic.replace(\"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/\", \"/\")} \\\n",
    "    {f_tfs.replace(\"/data/genome/drosophila_melanogaster/\", \"/\")} \\\n",
    "    -o {adj_matrix.replace(\"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/\", \"/\")} \\\n",
    "    --num_workers {sc.settings.njobs}\n",
    "\n",
    "# Note: It takes ~2h to complete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de57838f",
   "metadata": {},
   "source": [
    "Read in the adjacencies matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7da47991",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bi</td>\n",
       "      <td>lncRNA:CR32773</td>\n",
       "      <td>7.387206e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CG9650</td>\n",
       "      <td>lncRNA:CR44357</td>\n",
       "      <td>5.064458e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trv</td>\n",
       "      <td>CG34354</td>\n",
       "      <td>4.570765e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CG34354</td>\n",
       "      <td>trv</td>\n",
       "      <td>4.545700e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>br</td>\n",
       "      <td>Mur2B</td>\n",
       "      <td>4.179730e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4016712</th>\n",
       "      <td>Dgp-1</td>\n",
       "      <td>Hug</td>\n",
       "      <td>7.302217e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4016713</th>\n",
       "      <td>rump</td>\n",
       "      <td>CG8401</td>\n",
       "      <td>3.989110e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4016714</th>\n",
       "      <td>crol</td>\n",
       "      <td>Capa</td>\n",
       "      <td>3.451499e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4016715</th>\n",
       "      <td>pho</td>\n",
       "      <td>Ilp3</td>\n",
       "      <td>1.136389e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4016716</th>\n",
       "      <td>pho</td>\n",
       "      <td>Ilp2</td>\n",
       "      <td>3.013293e-21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4016717 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TF          target    importance\n",
       "0             bi  lncRNA:CR32773  7.387206e+02\n",
       "1         CG9650  lncRNA:CR44357  5.064458e+02\n",
       "2            trv         CG34354  4.570765e+02\n",
       "3        CG34354             trv  4.545700e+02\n",
       "4             br           Mur2B  4.179730e+02\n",
       "...          ...             ...           ...\n",
       "4016712    Dgp-1             Hug  7.302217e-20\n",
       "4016713     rump          CG8401  3.989110e-20\n",
       "4016714     crol            Capa  3.451499e-20\n",
       "4016715      pho            Ilp3  1.136389e-20\n",
       "4016716      pho            Ilp2  3.013293e-21\n",
       "\n",
       "[4016717 rows x 3 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adjacencies = pd.read_csv(adj_matrix, sep=',')\n",
    "#adjacencies.to_csv(adj_matrix.replace(\".csv\", \".tsv\"), index=False, sep='\\t')\n",
    "adjacencies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10e400d2",
   "metadata": {},
   "source": [
    "## STEP 2-3: Regulon prediction aka cisTarget from CLI\n",
    "\n",
    "For this step the CLI version of SCENIC is used. This step can be deployed on an High Performance Computing system.\n",
    "\n",
    "*Output:* List of adjacencies between a TF and its targets.\n",
    "\n",
    "### 2.a. Running regulon prediction using cisTarget\n",
    "\n",
    "Here, we use the --mask_dropouts option, which affects how the correlation between TF and target genes is calculated during module creation. It is important to note that prior to pySCENIC v0.9.18, the default behavior was to mask dropouts, while in v0.9.18 and later, the correlation is performed using the entire set of cells (including those with zero expression). When using the modules_from_adjacencies function directly in python instead of via the command line, the rho_mask_dropouts option can be used to control this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "dd5ac009",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:00:57,846 - pyscenic.cli.pyscenic - INFO - Creating modules.\n",
      "\n",
      "2024-03-06 19:00:59,393 - pyscenic.cli.pyscenic - INFO - Loading expression matrix.\n",
      "\n",
      "2024-03-06 19:01:21,750 - pyscenic.utils - INFO - Calculating Pearson correlations.\n",
      "\n",
      "2024-03-06 19:01:22,646 - pyscenic.utils - WARNING - Note on correlation calculation: the default behaviour for calculating the correlations has changed after pySCENIC verion 0.9.16. Previously, the default was to calculate the correlation between a TF and target gene using only cells with non-zero expression values (mask_dropouts=True). The current default is now to use all cells to match the behavior of the R verision of SCENIC. The original settings can be retained by setting 'rho_mask_dropouts=True' in the modules_from_adjacencies function, or '--mask_dropouts' from the CLI.\n",
      "\tDropout masking is currently set to [False].\n",
      "\n",
      "2024-03-06 19:03:25,003 - pyscenic.utils - INFO - Creating modules.\n",
      "\n",
      "2024-03-06 19:06:17,636 - pyscenic.cli.pyscenic - INFO - Loading databases.\n",
      "\n",
      "2024-03-06 19:06:17,724 - pyscenic.cli.pyscenic - INFO - Calculating regulons.\n",
      "\n",
      "2024-03-06 19:06:17,725 - pyscenic.prune - INFO - Using 112 workers.\n",
      "\n",
      "2024-03-06 19:06:17,725 - pyscenic.prune - INFO - Using 112 workers.\n",
      "\n",
      "2024-03-06 19:06:18,584 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,584 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,746 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,746 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,795 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,795 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,817 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,817 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,981 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:18,981 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,005 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,005 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,029 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,029 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,085 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,085 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,163 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,163 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,181 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,181 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,335 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,335 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,343 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,343 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,423 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,423 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,443 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,443 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,479 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,479 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,650 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,650 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,698 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,698 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): database loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,707 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,707 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,713 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,713 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): motif annotations loaded in memory.\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "\n",
      "2024-03-06 19:06:19,729 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,729 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): motif annotations loaded in memory.\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/venv/bin/pyscenic\", line 33, in <module>\n",
      "    sys.exit(load_entry_point('pyscenic==0.12.1+0.gce41b61.dirty', 'console_scripts', 'pyscenic')())\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/pyscenic/cli/pyscenic.py\", line 713, in main\n",
      "    args.func(args)\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/pyscenic/cli/pyscenic.py\", line 236, in prune_targets_command\n",
      "    df_motifs = calc_func(\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/pyscenic/prune.py\", line 424, in prune2df\n",
      "    return _distributed_calc(\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/pyscenic/prune.py\", line 245, in _distributed_calc\n",
      "    ).start()\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/multiprocessing_on_dill/process.py\", line 105, in start\n",
      "    self._popen = self._Popen(self)\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/multiprocessing_on_dill/context.py\", line 212, in _Popen\n",
      "    return _default_context.get_context().Process._Popen(process_obj)\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/multiprocessing_on_dill/context.py\", line 267, in _Popen\n",
      "    return Popen(process_obj)\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/multiprocessing_on_dill/popen_fork.py\", line 20, in __init__\n",
      "    self._launch(process_obj)\n",
      "  File \"/opt/venv/lib/python3.10/site-packages/multiprocessing_on_dill/popen_fork.py\", line 67, in _launch\n",
      "    self.pid = os.fork()\n",
      "BlockingIOError: [Errno 11] Resource temporarily unavailable\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:19,914 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,914 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,926 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,926 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): motif annotations loaded in memory.\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "\n",
      "2024-03-06 19:06:19,984 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,984 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,995 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:19,995 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,026 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,026 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): motif annotations loaded in memory.\n",
      "terminate called after throwing an instance of 'std::system_error'\n",
      "  what():  Resource temporarily unavailable\n",
      "\n",
      "2024-03-06 19:06:20,030 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,030 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,142 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,142 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,152 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,152 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,196 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,196 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,239 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,239 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,352 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,352 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,367 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:20,367 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): motif annotations loaded in memory.\n",
      "\n",
      "2024-03-06 19:06:22,625 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Eip78C could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:22,872 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG2931 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,007 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4546 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,134 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hr4 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,171 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG11504 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,199 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG2964 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,437 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4593 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,521 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for TfIIFalpha could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,528 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Prosalpha1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,566 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG8319 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,585 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Psi could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,615 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hr51 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,712 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ara could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,735 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for l(2)k10201 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,807 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Elba2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,863 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Tfb4 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,919 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Ptip could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,932 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG8578 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,942 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fru could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,949 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG11617 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,976 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG30178 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:23,992 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Abd-B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,014 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Mondo could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,042 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for arg could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,078 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Cyt-c-d could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,084 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4611 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:24,233 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hr83 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,234 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Ptx1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,242 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ftz could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,252 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9121 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,346 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Abl could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,398 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4730 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,409 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lab could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,501 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Ets21C could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,560 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Topors could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,570 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9246 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,578 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG11839 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,730 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Myb could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:24,785 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ase could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,178 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4854 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,278 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Sidpn could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,294 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Ets96B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,364 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9314 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,373 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG12071 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,406 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hrb87F could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,431 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31064 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,443 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for NC2alpha could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,488 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4935 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,535 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lbl could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,563 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ato could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,665 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Trax could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,692 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for gcm could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,793 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5023 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,799 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for DNApol-iota could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,853 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Antp could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,925 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Sirt6 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,953 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Trf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,973 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for gcm2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:25,991 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG12129 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,046 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for NELF-B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,048 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31388 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,120 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Dab could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,195 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Fer1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,219 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bab1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,242 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9609 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,247 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Six4 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,283 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for gem could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,359 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31612 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,406 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5116 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,437 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hsf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:26,473 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for NK7.1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,490 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lin-28 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,525 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bap could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,540 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9650 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,580 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Dad could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,604 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG12219 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,661 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31624 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,703 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5144 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,748 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for danr could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,773 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Fer3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,789 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lmd could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,821 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bbx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,865 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for glo could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,874 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Dbx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,991 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Irbp18 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:26,998 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5180 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,039 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dar1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,083 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lms could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,103 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG9876 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,139 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bi could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,142 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Asap could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,255 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31674 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,296 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5245 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,354 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for FoxK could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,444 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Asciz could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,559 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG31988 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,596 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5285 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,644 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for FoxL1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,751 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dimm could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,817 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bin could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,819 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for lov could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,829 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Dgp-1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,862 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG32075 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,965 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for FoxP could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,972 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CR43670 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:27,972 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Nf-YA could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,030 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for UQCR-14L could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,039 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Atf-2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,126 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bnl could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,148 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for luna could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,149 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG12768 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,154 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3251 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,179 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5728 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,241 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for GATAd could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,241 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CR4575 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:28,454 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG32532 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,615 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Ubx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,730 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for mRpL1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,741 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3294 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,758 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG5890 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,786 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bon could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,805 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Gem3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,824 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Dlip3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:28,886 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Nf1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,058 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dmrt11E could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,262 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for NfI could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,349 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dmrt93B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,452 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3347 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,512 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for br could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,620 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dmrt99B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,705 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Doa could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,715 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG14641 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,742 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG33722 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,767 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG6083 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,789 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Gsc could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:29,878 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Nup107 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,043 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Usp39 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,044 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dpn could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,137 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG14667 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,166 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG34031 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,209 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for H15 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,216 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Doc1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,240 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Nup133 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,327 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RIOK2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,427 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG14718 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,450 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3407 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,475 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG6388 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,575 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Doc2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,589 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dsf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,621 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Vsx1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,716 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG14860 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,758 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for HGTX could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,839 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for gsb could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,905 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for XRCC1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:30,995 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Kah could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,010 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG15011 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,040 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for HHEX could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,051 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG6769 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,059 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG34354 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,079 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for mil could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:31,140 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for OdsH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,149 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for bsh could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,300 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG15601 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,321 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for HLH3B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,336 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7069 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,346 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG34367 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,404 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for gt could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,434 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for btd could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,456 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for dysf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,511 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Xpac could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,556 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Kdm4A could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,590 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG15696 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,600 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for HLH4C could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,623 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Chrac-14 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,633 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3708 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,666 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for mio could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,670 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Atf6 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,682 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for mirr could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,722 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for byn could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,739 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for e(y)1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,840 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Kdm4B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,923 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3815 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:31,968 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Awh could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,015 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for cad could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,107 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rb97D could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,129 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Kr could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,173 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG15812 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,195 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7362 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,229 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CkIIalpha-i1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,289 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for PICK1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,309 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for caup could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,390 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rbbp5 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,483 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7368 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,498 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG3919 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,527 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CkIIbeta could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,593 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for eg could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,602 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for caz could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,675 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rbf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,696 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Sox15 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,791 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4038 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,809 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG17287 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:32,810 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Clamp could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,008 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for PTPMT1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,032 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nau could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:33,083 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rbf2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,119 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Sox21a could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,132 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Sox21b could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,140 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Zif could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,202 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7386 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,229 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG17440 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,270 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for hkb could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,276 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for E(spl)m5-HLH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,390 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Parp could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,437 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Lmx1a could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,441 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for abd-A could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,497 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7745 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,525 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG42741 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,526 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Clp could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,539 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for hng1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,715 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ac could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,718 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for E(spl)m7-HLH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,755 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rbp4 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,807 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for hng3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,815 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG17829 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,825 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4328 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,857 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for hppy could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:33,872 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Brf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,224 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Cnx14D could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,267 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG43347 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,271 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for achi could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,281 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Bro could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,293 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for en could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,312 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for E(spl)m8-HLH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,358 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hand could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,448 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Spps could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,518 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG18599 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,566 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7839 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,612 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ind could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,627 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for erm could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,641 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hcf could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,670 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for MTF-1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,814 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG1924 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,824 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RfC38 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,840 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for al could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,862 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG7878 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,865 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4424 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,893 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for inv could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,894 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for BuGZ could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:34,921 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hcs could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:34,951 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Mad could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,026 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Srp9 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,119 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for amos could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,121 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RfC4 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,175 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for jigr1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,180 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for eve could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,191 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for C15 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,198 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hesr could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,211 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG8079 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,376 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for croc could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,402 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG2120 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,423 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,423 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,424 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): Done.\n",
      "\n",
      "2024-03-06 19:06:35,424 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(12): Done.\n",
      "\n",
      "2024-03-06 19:06:35,459 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ewg could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,473 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hey could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,488 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG44422 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,512 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rfx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,556 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG8152 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,654 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,654 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,655 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): Done.\n",
      "\n",
      "2024-03-06 19:06:35,655 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(5): Done.\n",
      "\n",
      "2024-03-06 19:06:35,690 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG2608 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,784 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG4496 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,788 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,788 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,789 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): Done.\n",
      "\n",
      "2024-03-06 19:06:35,789 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(3): Done.\n",
      "\n",
      "2024-03-06 19:06:35,816 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rox8 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,865 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for jing could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:35,999 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:35,999 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:36,000 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): Done.\n",
      "\n",
      "2024-03-06 19:06:36,000 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(2): Done.\n",
      "\n",
      "2024-03-06 19:06:36,022 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Poxn could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,027 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for exex could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,139 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for jumu could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,184 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:36,184 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:36,185 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): Done.\n",
      "\n",
      "2024-03-06 19:06:36,185 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(4): Done.\n",
      "\n",
      "2024-03-06 19:06:36,313 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for PpD3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,316 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:36,316 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:36,318 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): Done.\n",
      "\n",
      "2024-03-06 19:06:36,318 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(9): Done.\n",
      "\n",
      "2024-03-06 19:06:36,319 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hmx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,610 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fd102C could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,652 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG10431 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,870 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fd19B could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,913 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG10463 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:36,916 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nerfin-1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:37,091 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RpS10a could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,257 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fd59A could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,310 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Mettl14 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,317 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nerfin-2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,359 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for E(spl)mdelta-HLH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,528 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fd96Ca could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,575 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Mettl3 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,590 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for net could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,650 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for E(spl)mgamma-HLH could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:37,784 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for fd96Cb could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:38,045 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:38,045 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:38,047 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): Done.\n",
      "\n",
      "2024-03-06 19:06:38,047 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(14): Done.\n",
      "\n",
      "2024-03-06 19:06:38,136 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG11085 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:38,146 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nom could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:38,583 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for CG11360 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:38,834 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rrp1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:38,863 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:38,863 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:38,869 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): Done.\n",
      "\n",
      "2024-03-06 19:06:38,869 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(1): Done.\n",
      "\n",
      "2024-03-06 19:06:38,913 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for ECSIT could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,035 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:39,035 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:39,037 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): Done.\n",
      "\n",
      "2024-03-06 19:06:39,037 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(13): Done.\n",
      "\n",
      "2024-03-06 19:06:39,119 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rrp40 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,124 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nos could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,398 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nub could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,403 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RunxA could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,427 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for TfAP-2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,660 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for nvy could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,663 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for RunxB could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,936 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Rx could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:39,937 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:39,937 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:39,939 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): Done.\n",
      "\n",
      "2024-03-06 19:06:39,939 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(16): Done.\n",
      "\n",
      "2024-03-06 19:06:40,196 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for kn could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:40,229 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for TfIIA-S-2 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:40,241 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,241 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,245 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): Done.\n",
      "\n",
      "2024-03-06 19:06:40,245 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(8): Done.\n",
      "\n",
      "2024-03-06 19:06:40,460 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for SF1 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:40,489 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,489 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,491 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): Done.\n",
      "\n",
      "2024-03-06 19:06:40,491 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(11): Done.\n",
      "\n",
      "2024-03-06 19:06:40,942 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Hr39 could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:40,946 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,946 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:40,949 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): Done.\n",
      "\n",
      "2024-03-06 19:06:40,949 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(7): Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2024-03-06 19:06:41,203 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for l(1)sc could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:41,432 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for l(2)37Cd could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:41,830 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for l(2)dtl could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:41,833 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:41,833 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:41,838 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): Done.\n",
      "\n",
      "2024-03-06 19:06:41,838 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(15): Done.\n",
      "\n",
      "2024-03-06 19:06:42,070 - pyscenic.transform - WARNING - Less than 80% of the genes in Regulon for Eip74EF could be mapped to dm6_v10_clust.genes_vs_motifs.rankings. Skipping this module.\n",
      "\n",
      "2024-03-06 19:06:42,378 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:42,378 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:42,380 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): Done.\n",
      "\n",
      "2024-03-06 19:06:42,380 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(6): Done.\n",
      "\n",
      "2024-03-06 19:06:42,763 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:42,763 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): All regulons derived.\n",
      "\n",
      "2024-03-06 19:06:42,770 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): Done.\n",
      "\n",
      "2024-03-06 19:06:42,770 - pyscenic.prune - INFO - Worker dm6_v10_clust.genes_vs_motifs.rankings(10): Done.\n"
     ]
    }
   ],
   "source": [
    "# Here I run the CLI version from the Docker\n",
    "!docker run -it --rm --name pySCENIC \\\n",
    "    -v /data/genome/drosophila_melanogaster/cistopic_flybase_r6.02:/cistopic_flybase_r6.02 \\\n",
    "    -v /home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/analysis:/analysis \\\n",
    "    -w /analysis aertslab/pyscenic:0.12.1 \\\n",
    "    pyscenic ctx \\\n",
    "    {adj_matrix.replace(\"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/\", \"/\")} \\\n",
    "    {f_db_names.replace(\"/data/genome/drosophila_melanogaster/\", \"/\")} \\\n",
    "    --annotations_fname {f_motif_path.replace(\"/data/genome/drosophila_melanogaster/\", \"/\")} \\\n",
    "    --expression_mtx_fname {f_loom_path_scenic.replace(\"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/\", \"/\")} \\\n",
    "    --output {reg_matrix.replace(\"/home/gardeux/SVRAW1/gardeux/2023-04-19_Emma_snRNAseq/\", \"/\")} \\\n",
    "    #--mask_dropouts \\\n",
    "    --num_workers 12\n",
    "            \n",
    "# Note: It takes ~2mn to complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c1c49c6e",
   "metadata": {},
   "outputs": [
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[45], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m regulons \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreg_matrix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m,\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m regulons\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/util/_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[0;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    947\u001b[0m )\n\u001b[1;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py:605\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    602\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    604\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 605\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1442\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1439\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1441\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1442\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1753\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1750\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1752\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1753\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1754\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1755\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py:79\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     76\u001b[0m     kwds\u001b[38;5;241m.\u001b[39mpop(key, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m     78\u001b[0m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m ensure_dtype_objs(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[0;32m---> 79\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m \u001b[43mparsers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     81\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munnamed_cols \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39munnamed_cols\n\u001b[1;32m     83\u001b[0m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:554\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "regulons = pd.read_csv(reg_matrix, sep=',')\n",
    "regulons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5c92eea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Enrichment</th>\n",
       "      <th>Enrichment.1</th>\n",
       "      <th>Enrichment.2</th>\n",
       "      <th>Enrichment.3</th>\n",
       "      <th>Enrichment.4</th>\n",
       "      <th>Enrichment.5</th>\n",
       "      <th>Enrichment.6</th>\n",
       "      <th>Enrichment.7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AUC</td>\n",
       "      <td>NES</td>\n",
       "      <td>MotifSimilarityQvalue</td>\n",
       "      <td>OrthologousIdentity</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>Context</td>\n",
       "      <td>TargetGenes</td>\n",
       "      <td>RankAtMax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TF</td>\n",
       "      <td>MotifID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Atf3</td>\n",
       "      <td>stark__TGANTCA</td>\n",
       "      <td>0.05681504924880987</td>\n",
       "      <td>4.666138672306601</td>\n",
       "      <td>3.99369e-06</td>\n",
       "      <td>0.068314</td>\n",
       "      <td>gene is orthologous to ENSG00000162772 in H. s...</td>\n",
       "      <td>frozenset({'weight&gt;75.0%', 'dm6_v10_clust.gene...</td>\n",
       "      <td>[('CG42748', 1.12564711607615), ('Gclm', 1.078...</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Atf3</td>\n",
       "      <td>metacluster_148.1</td>\n",
       "      <td>0.049568211457107415</td>\n",
       "      <td>3.173405446015911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.307087</td>\n",
       "      <td>motif is annotated for orthologous gene ENSG00...</td>\n",
       "      <td>frozenset({'weight&gt;75.0%', 'dm6_v10_clust.gene...</td>\n",
       "      <td>[('Ac13E', 0.7477660264325211), ('Dh31-R', 1.1...</td>\n",
       "      <td>4988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Atf3</td>\n",
       "      <td>metacluster_157.2</td>\n",
       "      <td>0.05377344487208033</td>\n",
       "      <td>4.039616588484887</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.384</td>\n",
       "      <td>motif is annotated for orthologous gene ENSG00...</td>\n",
       "      <td>frozenset({'weight&gt;75.0%', 'dm6_v10_clust.gene...</td>\n",
       "      <td>[('CG42784', 1.511065204057486), ('Fim', 3.418...</td>\n",
       "      <td>652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>srp</td>\n",
       "      <td>metacluster_117.4</td>\n",
       "      <td>0.06427245428554983</td>\n",
       "      <td>5.9103326489220915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'dm6_v10_clust.genes_...</td>\n",
       "      <td>[('Ac13E', 15.015438165840512), ('CG17646', 9....</td>\n",
       "      <td>4765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>srp</td>\n",
       "      <td>metacluster_117.7</td>\n",
       "      <td>0.05220225646142333</td>\n",
       "      <td>3.774395080426182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'dm6_v10_clust.genes_...</td>\n",
       "      <td>[('Idh', 9.23347691913659), ('mbc', 6.78963765...</td>\n",
       "      <td>4725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>srp</td>\n",
       "      <td>metacluster_169.2</td>\n",
       "      <td>0.0685249055010841</td>\n",
       "      <td>6.66284478184673</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'dm6_v10_clust.genes_...</td>\n",
       "      <td>[('dpp', 2.5538793133476654), ('for', 24.07611...</td>\n",
       "      <td>4996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>tj</td>\n",
       "      <td>tfdimers__MD00092</td>\n",
       "      <td>0.09257796501082441</td>\n",
       "      <td>3.113835207142027</td>\n",
       "      <td>2.71523e-07</td>\n",
       "      <td>0.391026</td>\n",
       "      <td>motif similar to tfdimers__MD00162 ('M01721_fo...</td>\n",
       "      <td>frozenset({'activating', 'dm6_v10_clust.genes_...</td>\n",
       "      <td>[('msi', 4.820953696086392), ('bru2', 4.166019...</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>tj</td>\n",
       "      <td>tfdimers__MD00028</td>\n",
       "      <td>0.09124685506991984</td>\n",
       "      <td>3.0104484824509816</td>\n",
       "      <td>5.04243e-11</td>\n",
       "      <td>0.212613</td>\n",
       "      <td>gene is orthologous to ENSG00000182759 in H. s...</td>\n",
       "      <td>frozenset({'activating', 'dm6_v10_clust.genes_...</td>\n",
       "      <td>[('shep', 1.754661006506003), ('para', 4.42259...</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0         Unnamed: 1            Enrichment        Enrichment.1  \\\n",
       "0          NaN                NaN                   AUC                 NES   \n",
       "1           TF            MotifID                   NaN                 NaN   \n",
       "2         Atf3     stark__TGANTCA   0.05681504924880987   4.666138672306601   \n",
       "3         Atf3  metacluster_148.1  0.049568211457107415   3.173405446015911   \n",
       "4         Atf3  metacluster_157.2   0.05377344487208033   4.039616588484887   \n",
       "..         ...                ...                   ...                 ...   \n",
       "279        srp  metacluster_117.4   0.06427245428554983  5.9103326489220915   \n",
       "280        srp  metacluster_117.7   0.05220225646142333   3.774395080426182   \n",
       "281        srp  metacluster_169.2    0.0685249055010841    6.66284478184673   \n",
       "282         tj  tfdimers__MD00092   0.09257796501082441   3.113835207142027   \n",
       "283         tj  tfdimers__MD00028   0.09124685506991984  3.0104484824509816   \n",
       "\n",
       "              Enrichment.2         Enrichment.3  \\\n",
       "0    MotifSimilarityQvalue  OrthologousIdentity   \n",
       "1                      NaN                  NaN   \n",
       "2              3.99369e-06             0.068314   \n",
       "3                      0.0             0.307087   \n",
       "4                      0.0                0.384   \n",
       "..                     ...                  ...   \n",
       "279                    0.0                  1.0   \n",
       "280                    0.0                  1.0   \n",
       "281                    0.0                  1.0   \n",
       "282            2.71523e-07             0.391026   \n",
       "283            5.04243e-11             0.212613   \n",
       "\n",
       "                                          Enrichment.4  \\\n",
       "0                                           Annotation   \n",
       "1                                                  NaN   \n",
       "2    gene is orthologous to ENSG00000162772 in H. s...   \n",
       "3    motif is annotated for orthologous gene ENSG00...   \n",
       "4    motif is annotated for orthologous gene ENSG00...   \n",
       "..                                                 ...   \n",
       "279                         gene is directly annotated   \n",
       "280                         gene is directly annotated   \n",
       "281                         gene is directly annotated   \n",
       "282  motif similar to tfdimers__MD00162 ('M01721_fo...   \n",
       "283  gene is orthologous to ENSG00000182759 in H. s...   \n",
       "\n",
       "                                          Enrichment.5  \\\n",
       "0                                              Context   \n",
       "1                                                  NaN   \n",
       "2    frozenset({'weight>75.0%', 'dm6_v10_clust.gene...   \n",
       "3    frozenset({'weight>75.0%', 'dm6_v10_clust.gene...   \n",
       "4    frozenset({'weight>75.0%', 'dm6_v10_clust.gene...   \n",
       "..                                                 ...   \n",
       "279  frozenset({'activating', 'dm6_v10_clust.genes_...   \n",
       "280  frozenset({'activating', 'dm6_v10_clust.genes_...   \n",
       "281  frozenset({'activating', 'dm6_v10_clust.genes_...   \n",
       "282  frozenset({'activating', 'dm6_v10_clust.genes_...   \n",
       "283  frozenset({'activating', 'dm6_v10_clust.genes_...   \n",
       "\n",
       "                                          Enrichment.6 Enrichment.7  \n",
       "0                                          TargetGenes    RankAtMax  \n",
       "1                                                  NaN          NaN  \n",
       "2    [('CG42748', 1.12564711607615), ('Gclm', 1.078...          457  \n",
       "3    [('Ac13E', 0.7477660264325211), ('Dh31-R', 1.1...         4988  \n",
       "4    [('CG42784', 1.511065204057486), ('Fim', 3.418...          652  \n",
       "..                                                 ...          ...  \n",
       "279  [('Ac13E', 15.015438165840512), ('CG17646', 9....         4765  \n",
       "280  [('Idh', 9.23347691913659), ('mbc', 6.78963765...         4725  \n",
       "281  [('dpp', 2.5538793133476654), ('for', 24.07611...         4996  \n",
       "282  [('msi', 4.820953696086392), ('bru2', 4.166019...          635  \n",
       "283  [('shep', 1.754661006506003), ('para', 4.42259...          296  \n",
       "\n",
       "[284 rows x 10 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulons = pd.read_csv(reg_matrix, sep=',')\n",
    "regulons"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python v3.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
