---
title: "Emma Pan Neuro (Control + ND75KD) - Preprocessing Kells AD data - REGION dataset"
author: "Vincent Gardeux"
date: "2025/04/23"
last_modified: "2025/04/23"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/gardeux/Neuro_Droso_ND75KD/")
#setwd("/data/gardeux/Neuro_Droso_ND75KD/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat)) # Seurat v5 is needed here
suppressPackageStartupMessages(library(BPCells)) # For handling big scRNAseq data in Seurat
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(ggplot2)) # For plotting
suppressPackageStartupMessages(library(ggpubr)) # For grid plotting
suppressPackageStartupMessages(library(plotly)) # For interactive plots
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("SeuratObject"), "version", as.character(packageVersion("SeuratObject")), "\n")
cat(bold("BPCells"), "version", as.character(packageVersion("BPCells")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("ggpubr"), "version", as.character(packageVersion("ggpubr")), "\n")
cat(bold("plotly"), "version", as.character(packageVersion("plotly")), "\n")

# Random seed
set.seed(42)
```

## Parameters

*Paper:* https://www.nature.com/articles/s41586-024-07606-7

*Data repo:* Synapse

```{r}
## Parameters
input_seurat_object <- "./data/Kellis_2024/Gene Expression (snRNAseq - 10x) processed, multi-region/all_brain_regions_filt_preprocessed_scanpy_fullmatrix.rds"
input_Oxphos_119_AUCell_values_path <- "./data/Kellis_119_Oxphos_AUCell_auc.tsv.gz"
input_UPR_92_AUCell_values_path <- "./data/Kellis_92_UPR_AUCell_auc.tsv.gz"
output_figure_prefix <- "./figures/Figure_5A/Kellis_"
output_seurat_object <- "./data/Kellis_2024/Gene Expression (snRNAseq - 10x) processed, multi-region/all_brain_regions_filt_preprocessed_scanpy_fullmatrix_excitatory.neurons.only.rds"
```

## Functions

```{r}
plot.boxplot.aucell <- function(data.metadata, auc.name, metadata.name, pwidth = 5, pheight = 7, filename = NULL) {
  # Remove rows with NA in the metadata column
  boxplot_data <- data.metadata[!is.na(data.metadata[[metadata.name]]), ]

  # Calculate median values of AUC per group
  median_values <- tapply(boxplot_data[[auc.name]], boxplot_data[[metadata.name]], median)

  # Reorder factor levels of metadata based on median AUC
  boxplot_data[[metadata.name]] <- factor(boxplot_data[[metadata.name]], levels = names(sort(median_values, decreasing = FALSE)))

  # Plot
  p <- ggplot(boxplot_data, aes(x = .data[[metadata.name]], y = .data[[auc.name]])) +
    geom_boxplot(fill = "lightgrey") +
    labs(x = metadata.name, y = auc.name, title = metadata.name) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "none")

  # Save plots
  if(is.null(filename)) filename <- paste0(output_figure_prefix, auc.name, "_AUCell_ordered_boxplot_", metadata.name)
  ggsave(paste0(filename, ".pdf"), plot = p, width = pwidth, height = pheight, bg = "white")
  ggsave(paste0(filename, ".png"), plot = p, width = pwidth, height = pheight, dpi = 1000, bg = "white")

  return(p)
}
```

```{r}
plot.boxplot.aucell.2 <- function(data.metadata, auc.name, metadata.name, metadata.2.name, pwidth = 5, pheight = 7, filename = NULL) {
  # Remove rows with NA in the relevant metadata columns
  boxplot_data <- data.metadata[!is.na(data.metadata[[metadata.name]]) & !is.na(data.metadata[[metadata.2.name]]), ]

  # Calculate median AUC per group (based only on the main grouping)
  median_values <- tapply(boxplot_data[[auc.name]], boxplot_data[[metadata.name]], median)

  # Reorder factor levels of main grouping
  boxplot_data[[metadata.name]] <- factor(boxplot_data[[metadata.name]], levels = names(sort(median_values, decreasing = FALSE)))
  
  # Ensure the fill variable is a factor
  boxplot_data[[metadata.2.name]] <- as.factor(boxplot_data[[metadata.2.name]])

  # Plot with grouping
  p <- ggplot(boxplot_data, aes(x = .data[[metadata.name]], y = .data[[auc.name]], fill = .data[[metadata.2.name]])) +
    geom_boxplot(position = position_dodge(width = 0.8), outlier.size = 0.5) +
    labs(x = metadata.name, y = auc.name, fill = metadata.2.name, title = metadata.name) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

  # Save plots
  if (is.null(filename)) filename <- paste0(output_figure_prefix, auc.name, "_AUCell_ordered_grouped_boxplot_", metadata.name, "_by_", metadata.2.name)
  ggsave(paste0(filename, ".pdf"), plot = p, width = pwidth, height = pheight, bg = "white")
  ggsave(paste0(filename, ".png"), plot = p, width = pwidth, height = pheight, dpi = 1000, bg = "white")

  return(p)
}

```

## I. Loading Seurat object


### I.1 Read previously generated Seurat object (1.6M cells)

```{r}
data.seurat <- readRDS(input_seurat_object)
data.seurat
```

```{r}
p <- ggplot(data.seurat@meta.data, aes(x = U1, y = U2, color = region)) + geom_point()
p
```

### I.2 Reading and mapping clinical metadata


#### I.2.1 (To be run once) Fixing missing metadata

```{r}
data.clinical <- fread("./data/Kellis_2024/Supplementary_Table_1_sample_metadata.tsv", data.table = F)
data.clinical
```

```{r}
data.mapping <- fread("./data/Kellis_2024/mapping.txt", data.table = F)
data.mapping
```

```{r}
projid.table <- table(data.seurat@meta.data$projid)
rind.table <- table(data.seurat@meta.data$rind)
for(i in 1:nrow(data.mapping)){
    mapped.projid <- names(projid.table)[projid.table == data.mapping[i, "nbcells"]]
    data.clinical[data.clinical$subject == data.mapping[i, "Individual"], "projid"] <- mapped.projid
    for(region in c("AG", "EC", "HC", "MT", "PFC", "TH")){
        mapped.rind <- na.omit(unique(data.seurat@meta.data[data.seurat@meta.data$region == region & data.seurat@meta.data$projid == mapped.projid, "rind"]))
        if(length(mapped.rind) == 1){
            data.clinical[data.clinical$subject == data.mapping[i, "Individual"] & data.clinical$region == region, "rind"] <- mapped.rind
        } else {
            message("ERROR i=", i, " & projid=", mapped.projid, " & rind=", mapped.rind)
        }
    }
}
```

```{r}
fwrite(data.clinical, file = "./data/Kellis_2024/Supplementary_Table_1_sample_metadata.completed.tsv", quote = F, row.names = F, col.names = T, sep = "\t")
```

#### I.2.2 Reading clinical metadata from file

```{r}
data.clinical <- fread("./data/Kellis_2024/Supplementary_Table_1_sample_metadata.completed.tsv", data.table = F)
data.clinical
```

### I.3 Adding to Seurat object

```{r}
data.seurat$subject <- data.clinical$subject[match(data.seurat$projid, data.clinical$projid)]
data.seurat$msex <- data.clinical$msex[match(data.seurat$projid, data.clinical$projid)]
data.seurat$msex <- factor(data.seurat$msex, levels = as.character(0:1))
data.seurat$age_death <- data.clinical$age_death[match(data.seurat$projid, data.clinical$projid)]
data.seurat$age_death <- factor(data.seurat$age_death, levels = c("(75,80]", "(80,85]", "(85,90]", "90+"))
data.seurat$pmi <- data.clinical$pmi[match(data.seurat$projid, data.clinical$projid)]
data.seurat$pmi <- factor(data.seurat$pmi, levels = as.character(3:12))
data.seurat$pathAD <- data.clinical$pathAD[match(data.seurat$projid, data.clinical$projid)]
data.seurat$clinicalDiag <- data.clinical$clinicalDiag[match(data.seurat$projid, data.clinical$projid)]
```

## II.1 Reading OXPHOS AUC values from AUCell (using the 133 OXPHOS genes from KEGG Oxphos pathway https://www.genome.jp/entry/mmu00190)

```{r}
data.auc.oxphos <- fread(input_Oxphos_119_AUCell_values_path, data.table = F)
data.auc.oxphos
```

QC

```{r}
all(colnames(data.seurat) %in% data.auc.oxphos$Cell) # Number differs for unknown reason... It matches the Seurat objects though...
```

## II.2 Adding AUC metadata

```{r}
data.seurat$AUC_OXPHOS_119_KEGG <- data.auc.oxphos[match(colnames(data.seurat), data.auc.oxphos$Cell, nomatch = NA), "AUC"]
hist(data.seurat$AUC_OXPHOS_119_KEGG, breaks = 100)
```

## III.1 Reading Unfolded Protein Response (UPR) AUC values from AUCell

```{r}
data.auc.upr <- fread(input_UPR_92_AUCell_values_path, data.table = F)
data.auc.upr
```

QC

```{r}
all(colnames(data.seurat) %in% data.auc.upr$Cell) # Number differs for unknown reason... It matches the Seurat objects though...
```

## III.2 Adding AUC metadata

```{r}
data.seurat$AUC_UPR_92 <- data.auc.upr[match(colnames(data.seurat), data.auc.upr$Cell, nomatch = NA), "AUC"]
hist(data.seurat$AUC_UPR_92, breaks = 100)
```

## IV Boxplots

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "region")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "major.celltype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "minor.celltype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "full.exttype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype")
```

```{r fig.width=10, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", 10, 7)
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "hsubclass", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "age_death", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "pmi", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "pathAD", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "clinicalDiag", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "msex", 5, 7)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", "msex", pwidth = 10)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", "clinicalDiag", pwidth = 10)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", "pathAD", pwidth = 10)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", "age_death", pwidth = 15)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", "pmi", pwidth = 20)
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "region")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "major.celltype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "minor.celltype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "full.exttype")
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype")
```

```{r fig.width=10, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "cell_type_high_resolution", 10, 7)
```

```{r fig.width=5, fig.height=7}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "hsubclass", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "age_death", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "pmi", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "pathAD", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "clinicalDiag", 5, 7)
```

```{r}
plot.boxplot.aucell(data.seurat@meta.data, "AUC_UPR_92", "msex", 5, 7)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype", "msex", pwidth = 10)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype", "clinicalDiag", pwidth = 10)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype", "pathAD", pwidth = 10)
```

```{r}
sum(data.seurat$pathAD == "AD" & data.seurat$neuronal.exttype == "NRGN", na.rm = T)
sum(data.seurat$pathAD == "non-AD" & data.seurat$neuronal.exttype == "NRGN", na.rm = T)
sum(data.seurat$pathAD == "AD", na.rm = T)
sum(data.seurat$pathAD == "non-AD", na.rm = T)
```


```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype", "pmi", pwidth = 20)
```

```{r}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "neuronal.exttype", "age_death", pwidth = 15)
```

## Excitatory neurons only

```{r}
data.seurat.exc <- data.seurat[,data.seurat@meta.data$hcelltype == "Exc"]
data.seurat.exc
```

### Run the default pipeline (not correcting for batch effects, or integrating batches together since I don't know what are the batches? Is it projid?)

```{r}
data.seurat.exc <- NormalizeData(data.seurat.exc, normalization.method = "LogNormalize", scale.factor = 10000)
data.seurat.exc <- FindVariableFeatures(data.seurat.exc, selection.method = "vst", nfeatures = 2000)
data.seurat.exc <- ScaleData(data.seurat.exc, features = rownames(data.seurat.exc))
data.seurat.exc <- RunPCA(data.seurat.exc, features = VariableFeatures(object = data.seurat.exc), npcs = 150)
```

```{r}
ElbowPlot(data.seurat.exc, ndims = 120)
```

```{r}
# Jackstraw not optimized for dense 
#data.seurat.exc <- JackStraw(data.seurat.exc, num.replicate = 100, dims = 150)
#data.seurat.exc <- ScoreJackStraw(data.seurat.exc, dims = 1:150)
#JackStrawPlot(data.seurat.exc, dims = 75:95)
```

```{r}
nb_PCs <- 100 # A bit arbitrary, Elbow plot sucks for big datasets, but the Jackstraw cannot be run

options(future.globals.maxSize = 8000 * 1024^2) # Needed for FindNeighbors not to crash
data.seurat.exc <- FindNeighbors(data.seurat.exc, dims = 1:nb_PCs)
data.seurat.exc <- FindClusters(data.seurat.exc, resolution = 0.5)
data.seurat.exc <- RunUMAP(data.seurat.exc, dims = 1:nb_PCs)
data.seurat.exc <- RunTSNE(data.seurat.exc, dims = 1:nb_PCs)
```

### Save Seurat object

```{r}
saveRDS(data.seurat.exc, file = output_seurat_object)
```

```{r}
data.seurat.exc <- readRDS(output_seurat_object)
```

```{r}
data.seurat.exc$msex <- factor(data.seurat.exc$msex, levels = as.character(0:1))
data.seurat.exc$age_death <- factor(data.seurat.exc$age_death, levels = c("(75,80]", "(80,85]", "(85,90]", "90+"))
data.seurat.exc$pmi <- factor(data.seurat.exc$pmi, levels = as.character(3:12))
```

### Boxplots

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_cell_type_high_resolution_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_cell_type_high_resolution_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "neuronal.exttype", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_neuronal.exttype_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "neuronal.exttype", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_neuronal.exttype_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "msex", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_msex_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$msex == "0"]
b <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$msex == "1"]
t.test(a,b)
wilcox.test(a,b)
message("Mean msex_0 = ", mean(a))
message("Mean msex_1 = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "msex", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_msex_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$msex == "0"]
b <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$msex == "1"]
t.test(a,b)
wilcox.test(a,b)
message("Mean msex_0 = ", mean(a))
message("Mean msex_1 = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "clinicalDiag", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_clinicalDiag_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$clinicalDiag == "no dementia"]
b <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$clinicalDiag == "AD dementia"]
t.test(a,b)
wilcox.test(a,b)
message("Mean clinicalDiag_no_dementia = ", mean(a))
message("Mean clinicalDiag_AD_dementia = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "clinicalDiag", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_clinicalDiag_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$clinicalDiag == "no dementia"]
b <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$clinicalDiag == "AD dementia"]
t.test(a,b)
wilcox.test(a,b)
message("Mean clinicalDiag_no_dementia = ", mean(a))
message("Mean clinicalDiag_AD_dementia = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "pathAD", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_pathAD_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$pathAD == "non-AD"]
b <- data.seurat.exc$AUC_OXPHOS_119_KEGG[data.seurat.exc$pathAD == "AD"]
t.test(a,b)
wilcox.test(a,b)
message("Mean pathAD_non-AD = ", mean(a))
message("Mean pathAD_AD = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "pathAD", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_pathAD_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$pathAD == "non-AD"]
b <- data.seurat.exc$AUC_UPR_92[data.seurat.exc$pathAD == "AD"]
t.test(a,b)
wilcox.test(a,b)
message("Mean pathAD_non-AD = ", mean(a))
message("Mean pathAD_AD = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "pmi", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_pmi_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "pmi", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_pmi_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "age_death", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_age_death_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "AUC_UPR_92", "age_death", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_age_death_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", "msex", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_msex_by_cell_type_high_resolution_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", "clinicalDiag", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_clinicalDiag_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", "pathAD", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_pathAD_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", "pmi", pwidth = 20, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_pmi_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_OXPHOS_119_KEGG", "cell_type_high_resolution", "age_death", pwidth = 15, filename = paste0(output_figure_prefix, "AUC_OXPHOS_119_KEGG_AUCell_ordered_boxplot_age_death_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "msex", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_msex_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "clinicalDiag", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_clinicalDiag_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "pathAD", pwidth = 10, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_pathAD_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "pmi", pwidth = 20, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_pmi_by_cell_type_high_resolution_Excitatory.neurons-only"))

```

```{r}
plot.boxplot.aucell.2(data.seurat.exc@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "age_death", pwidth = 15, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_age_death_by_cell_type_high_resolution_Excitatory.neurons-only"))
```

### UMAP

```{r fig.width=10, fig.height=10}
p <- DimPlot(data.seurat.exc, reduction = "umap", label = T, group.by = "cell_type_high_resolution", raster = F) + NoLegend()
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_cell_type_high_resolution.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_cell_type_high_resolution.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- DimPlot(data.seurat.exc, reduction = "umap", label = T, group.by = "hcluster", raster = F) + NoLegend()
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_hcluster.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_hcluster.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- DimPlot(data.seurat.exc, reduction = "umap", label = T, group.by = "neuronal.layer", raster = F) + NoLegend()
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_neuronal.layer.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_neuronal.layer.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- DimPlot(data.seurat.exc, reduction = "umap", label = T, group.by = "neuronal.exttype", raster = F) + NoLegend()
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_neuronal.exttype.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_neuronal.exttype.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```
```{r fig.width=10, fig.height=10}
p <- DimPlot(data.seurat.exc, reduction = "umap", label = T, group.by = "region", raster = F) + NoLegend()
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_region.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_region.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- FeaturePlot(data.seurat.exc, reduction = "umap", features = "AUC_OXPHOS_119_KEGG", raster = F)
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_AUC_OXPHOS_119_KEGG.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_AUC_OXPHOS_119_KEGG.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- FeaturePlot(data.seurat.exc, reduction = "umap", features = "AUC_UPR_92", raster = F)
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_AUC_UPR_92.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_AUC_UPR_92.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

### ATF4 regulon

Import results from pySCENIC pipeline

```{r}
data.regulon <- fread("./data/Kellis_2024/Kellis_region_regulons_aucell.tsv", sep = "\t", data.table = F)
rownames(data.regulon) <- data.regulon$Cell
data.regulon.binarized <- fread("./data/Kellis_2024/Kellis_region_regulons_aucell_binarized.tsv", sep = "\t", data.table = F)
rownames(data.regulon.binarized) <- data.regulon.binarized$Cell
```

Add to Seurat object

```{r}
data.seurat.exc$ATF4_regulon <- data.regulon[colnames(data.seurat.exc), "ATF4(+)"]
data.seurat.exc$ATF4_regulon_binarized <- data.regulon.binarized[colnames(data.seurat.exc), "ATF4(+)"]
```

### Boxplots

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "cell_type_high_resolution", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_cell_type_high_resolution_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "neuronal.exttype", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_neuronal.exttype_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "msex", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_msex_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$ATF4_regulon[data.seurat.exc$msex == "0"]
b <- data.seurat.exc$ATF4_regulon[data.seurat.exc$msex == "1"]
t.test(a,b)
wilcox.test(a,b)
message("Mean msex_0 = ", mean(a))
message("Mean msex_1 = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "clinicalDiag", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_clinicalDiag_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$ATF4_regulon[data.seurat.exc$clinicalDiag == "no dementia"]
b <- data.seurat.exc$ATF4_regulon[data.seurat.exc$clinicalDiag == "AD dementia"]
t.test(a,b)
wilcox.test(a,b)
message("Mean clinicalDiag_no_dementia = ", mean(a))
message("Mean clinicalDiag_AD_dementia = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "pathAD", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_pathAD_Excitatory.neurons-only"))
```

```{r}
a <- data.seurat.exc$ATF4_regulon[data.seurat.exc$pathAD == "non-AD"]
b <- data.seurat.exc$ATF4_regulon[data.seurat.exc$pathAD == "AD"]
t.test(a,b)
wilcox.test(a,b)
message("Mean clinicalDiag_nn-AD = ", mean(a))
message("Mean clinicalDiag_AD = ", mean(b))
message("Fold-Change = ", mean(b) / mean(a))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "pmi", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_pmi_Excitatory.neurons-only"))
```

```{r}
plot.boxplot.aucell(data.seurat.exc@meta.data, "ATF4_regulon", "age_death", pwidth = 5, pheight = 7, filename = paste0(output_figure_prefix, "ATF4_regulon_AUCell_ordered_boxplot_age_death_Excitatory.neurons-only"))
```

### UMAPs

```{r fig.width=10, fig.height=10}
p <- FeaturePlot(data.seurat.exc, reduction = "umap", features = "ATF4_regulon", raster = F)
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_ATF4_regulon.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_ATF4_regulon.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```

```{r fig.width=10, fig.height=10}
p <- FeaturePlot(data.seurat.exc, reduction = "umap", features = "ATF4_regulon_binarized", raster = F)
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_ATF4_regulon_binarized.pdf"), plot = p, width = 10, height = 10, bg = "white")
ggsave(paste0(output_figure_prefix, "UMAP_Excitatory.neurons.only_ATF4_regulon_binarized.png"), plot = p, width = 10, height = 10, dpi = 1000, bg = "white")
p
```
