---
title: "Creating ASAP/SCENIC-compatible Loom file from Seurat object"
author: "Vincent Gardeux"
date: "2024/03/05"
updated: "2025/01/09"
output:
  html_document:
    df_print: paged
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/gardeux/Neuro_Droso_ND75KD/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat, lib.loc = "/software/Seuratv4/lib/")) # Seurat v4.4.0, library for single-cell analysis
suppressPackageStartupMessages(library(SeuratObject, lib.loc = "/software/Seuratv4/lib/")) # For compatibility purpose
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(loomR)) # For building a file for ASAP
suppressPackageStartupMessages(library(Matrix)) # For building a file for ASAP
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("SeuratObject"), "version", as.character(packageVersion("SeuratObject")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("loomR"), "version", as.character(packageVersion("loomR")), "\n")
cat(bold("Matrix"), "version", as.character(packageVersion("Matrix")), "\n")
```

## Parameters

```{r}
# Parameters
seurat_input <- "data/Pan_neuro_integrated_FINAL.rds"
loom_output <- "data/Pan_neuro_integrated_FINAL.loom"

# Random seed
set.seed(42)
```

## I.1 Reading Seurat object

First step is to read the Seurat object, previously created

```{r}
message("Loading Seurat object...")
data.seurat <- readRDS(seurat_input)
message(ncol(data.seurat), " cells were loaded")
message(nrow(data.seurat), " genes were loaded")
```

## I.2 Loading gene mapping from CellRanger output (I did not perform the CellRanger alignment)

```{r}
gene.mapping = fread('data/features.tsv', header = F, data.table = F)
colnames(gene.mapping) <- c("Ensembl", "Name", "Biotype")
gene.mapping$Name <- gsub(x = gene.mapping$Name, pattern = "_", replacement = "-")
rownames(gene.mapping) <- gene.mapping$Name
gene.mapping
```

## II. ASAP

If you want to use ASAP, you need to package the Seurat object into a Loom file, to upload on ASAP.

NOTE: I could use SeuratDisk SaveDisk option, but it does not do what I want. And there are many missing fields. So I'll create the Loom from scratch.

# Prepare Cell attributes to put in Loom

```{r}
attributes_cells <- list(CellID=data.seurat@assays$RNA@counts@Dimnames[[2]])
for(m in names(data.seurat@meta.data)){
  attributes_cells[[m]] <- data.seurat@meta.data[m]
}
```

# Main /matrix (I remove colnames and rownames)

```{r}
count_matrix <- data.seurat@assays$RNA@counts
colnames(count_matrix) <- NULL
rownames(count_matrix) <- NULL
```

# Create Loom file

```{r}
data.loom <- create(filename = loom_output, data = count_matrix, gene.attrs = list(Accession=gene.mapping[data.seurat@assays$RNA@counts@Dimnames[[1]],"Ensembl"], Gene=data.seurat@assays$RNA@counts@Dimnames[[1]]), cell.attrs = attributes_cells, overwrite = T)
```

# Adding Integrated NN graph

Generate b vector, from p

```{r}
b_vec <- rep(0, length(data.seurat@graphs$RNA_nn_harmony@i))
val <- 0
ind <- 0
for(i in 2:length(data.seurat@graphs$RNA_nn_harmony@p)){
  b_vec[(ind + 1):data.seurat@graphs$RNA_nn_harmony@p[i]] <- val
  val <- val + 1
  ind <- data.seurat@graphs$RNA_nn_harmony@p[i]
}
data.loom$add.graph(name = "RNA_nn_harmony", a = data.seurat@graphs$RNA_nn_harmony@i, b = b_vec, w = data.seurat@graphs$RNA_nn_harmony@x)
```

# Adding Integrated SNN graph

Generate b vector, from p

```{r}
b_vec <- rep(0, length(data.seurat@graphs$RNA_snn_harmony@i))
val <- 0
ind <- 0
for(i in 2:length(data.seurat@graphs$RNA_snn_harmony@p)){
  b_vec[(ind + 1):data.seurat@graphs$RNA_snn_harmony@p[i]] <- val
  val <- val + 1
  ind <- data.seurat@graphs$RNA_snn_harmony@p[i]
}
data.loom$add.graph(name = "RNA_snn_harmony", a = data.seurat@graphs$RNA_snn_harmony@i, b = b_vec, w = data.seurat@graphs$RNA_snn_harmony@x)
```

# Adding Layers

Saving also the normalized data in a layer

```{r}
data.loom$add.layer(layers = list(normalized=data.seurat@assays$RNA@data))
```

# Add Embeddings

Saving the PCA, UMAP and t-SNE coordinates

```{r}
data.loom$add.col.attribute(list(PCA=data.seurat@reductions$pca@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(t_SNE_uncorrected=data.seurat@reductions$tsne_uncorrected@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(UMAP_uncorrected=data.seurat@reductions$umap_uncorrected@cell.embeddings), overwrite = TRUE)

data.loom$add.col.attribute(list(Harmony=data.seurat@reductions$harmony@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(t_SNE_harmony=data.seurat@reductions$tsne_harmony@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(UMAP_harmony=data.seurat@reductions$umap_harmony@cell.embeddings), overwrite = TRUE)

data.loom$add.col.attribute(list(Harmony_overcorrected=data.seurat@reductions$harmony_overcorrected@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(t_SNE_harmony_overcorrected=data.seurat@reductions$tsne_harmony_overcorrected@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(UMAP_harmony_overcorrected=data.seurat@reductions$umap_harmony_overcorrected@cell.embeddings), overwrite = TRUE)

data.loom$add.col.attribute(list(SCENIC=data.seurat@reductions$SCENIC@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(t_SNE_SCENIC=data.seurat@reductions$SCENICtsne@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(UMAP_SCENIC=data.seurat@reductions$SCENICumap@cell.embeddings), overwrite = TRUE)

data.loom$add.col.attribute(list(SCENIC_binarized=data.seurat@reductions$SCENICbinarized@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(t_SNE_SCENIC_binarized=data.seurat@reductions$SCENICbinarizedtsne@cell.embeddings), overwrite = TRUE)
data.loom$add.col.attribute(list(UMAP_SCENIC_binarized=data.seurat@reductions$SCENICbinarizedumap@cell.embeddings), overwrite = TRUE)
```

# Close the Loom file

```{r}
data.loom$close_all()
```
