# Neuro_Droso_ND75-KD
Material and source code for the Neuro_Droso_ND75-KD manuscript

## 0. Raw sequencing data
All .fastq files were deposited on ArrayExpress under the accession number [E-MTAB-XXXXX](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-XXXXX). It comprises two 10x libraries (Pan_neuro_control and Pan_neuro_ND75KD).<br/>

### 0.1 Running CellRanger on the 10x single-nuclei libraries
Here we follow standard protocols. 

```bash
cellranger count --id Pan_neuro_control --fastqs=./fastq --sample=Pan_neuro_control --transcriptome=${10x_genome} --expect-cells=10000 --chemistry=auto --include-introns=true
cellranger count --id Pan_neuro_ND75KD --fastqs=./fastq --sample=Pan_neuro_ND75KD --transcriptome=${10x_genome} --expect-cells=10000 --chemistry=auto --include-introns=true
```
**Note:** In the manuscript, we used the dm6 genome assembly.

## 1. Preprocessing of the cellranger outputs

### 1.0. Generating individual Seurat object for each library <sub>(see code & output here: Pan_neuro_control.[[Rmd](preprocessing/S0a_Pan_neuro_control.Rmd)] and Pan_neuro_ND75KD.[[Rmd](preprocessing/S0b_Pan_neuro_ND75KD.Rmd)])</sub>
Here we load the .h5 files generated by cellranger in the previous step, filter the bad quality cells, and run standard Seurat v4.4.0 pipeline.

```bash
Rscript -e "rmarkdown::render('./preprocessing/S0a_Pan_neuro_control.Rmd', output_file = './preprocessing/S0a_Pan_neuro_control.html')"
Rscript -e "rmarkdown::render('./preprocessing/S0b_Pan_neuro_ND75KD.Rmd', output_file = './preprocessing/S0b_Pan_neuro_ND75KD.html')"
```

### 1.1. Integration of transcriptomics data with Harmony <sub>(see full code here: [[Rmd](preprocessing/S1_Pan_neuro_integrated.Rmd)])</sub>
Here we load the two previously generated Seurat objects, and we integrate them using Harmony. We both use a lenient integration method (with `early_stop = T`) and a "over-corrected" integration (with `early_stop = F` and `max_iter = 100`) to see where the new ND75KD-specific clusters are merged with. Then we run the standard Seurat v4.4.0 pipeline to generate UMAP, t-SNE and clustering metadata on the integrated Seurat object.

```bash
Rscript -e "rmarkdown::render('./preprocessing/S1_Pan_neuro_integrated.Rmd', output_file = './preprocessing/S1_Pan_neuro_integrated.html')"
```

### 1.2. Adding metadata to the Seurat object
Now that we generated the integrated Seurat object, we will add the following metadata:
a) An Oxphos score generated using AUCell from the pySCENIC pipeline
b) A manual annotation of the clusters

#### 1.2.a Oxphos score using AUCell
TODO

#### 1.2.b Manual annotation of the clusters
TODO

### 1.3. Running pySCENIC and adding regulon data to the Seurat object
Now, we will run the pySCENIC pipeline in Python, using the aertslab/pyscenic:0.12.1 Docker image available on the DockerHub.
Of note: We simply had to add the `ipykernel` package in the Docker image to create a Jupyter Notebook kernel (see 

## 2. Manuscript Figures

All manuscript figures are reproducible through scripts deposited [here](./figures/)
