---
project: "Emma Pan Neuro (Control + ND75KD)"
author: "Vincent Gardeux"
date: "2025/06/05"
last_modified: "2025/06/05"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/gardeux/Neuro_Droso_ND75KD/")
#setwd("/data/gardeux/Neuro_Droso_ND75KD/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat)) # Seurat v5 is needed here
suppressPackageStartupMessages(library(BPCells)) # For handling big scRNAseq data in Seurat
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(ggplot2)) # For plotting
suppressPackageStartupMessages(library(ggpubr)) # For grid plotting
suppressPackageStartupMessages(library(plotly)) # For interactive plots
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("SeuratObject"), "version", as.character(packageVersion("SeuratObject")), "\n")
cat(bold("BPCells"), "version", as.character(packageVersion("BPCells")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("ggpubr"), "version", as.character(packageVersion("ggpubr")), "\n")
cat(bold("plotly"), "version", as.character(packageVersion("plotly")), "\n")

# Random seed
set.seed(42)
```

## Parameters

*Paper:*
Unravelling cell type-specific responses to Parkinsonâ€™s Disease at single cell resolution
Martirosyan et al.
Mol Neurodegeneration 19, 7 (2024)
https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-023-00699-0

*Data repo:* https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE243639

```{r}
## Parameters
input_seurat_object <- "./data/Martirosyan_2024/Martirosyan_neurons.rds"
output_figure_prefix <- "./figures/Figure_6E/Martirosyan_neurons_"
```

## Functions

```{r}
plot.boxplot.aucell <- function(data.metadata, auc.name, metadata.name, pwidth = 5, pheight = 7, filename = NULL) {
  # Remove rows with NA in the metadata column
  boxplot_data <- data.metadata[!is.na(data.metadata[[metadata.name]]), ]

  # Calculate median values of AUC per group
  median_values <- tapply(boxplot_data[[auc.name]], boxplot_data[[metadata.name]], median)

  # Reorder factor levels of metadata based on median AUC
  boxplot_data[[metadata.name]] <- factor(boxplot_data[[metadata.name]], levels = names(sort(median_values, decreasing = FALSE)))

  # Plot
  p <- ggplot(boxplot_data, aes(x = .data[[metadata.name]], y = .data[[auc.name]])) +
    geom_boxplot(fill = "lightgrey") +
    labs(x = metadata.name, y = auc.name, title = metadata.name) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "none")

  # Save plots
  if(is.null(filename)) filename <- paste0(output_figure_prefix, auc.name, "_AUCell_ordered_boxplot_", metadata.name)
  ggsave(paste0(filename, ".pdf"), plot = p, width = pwidth, height = pheight, bg = "white")
  ggsave(paste0(filename, ".png"), plot = p, width = pwidth, height = pheight, dpi = 1000, bg = "white")

  return(p)
}
```

```{r}
plot.boxplot.aucell.2 <- function(data.metadata, auc.name, metadata.name, metadata.2.name, pwidth = 5, pheight = 7, filename = NULL) {
  # Remove rows with NA in the relevant metadata columns
  boxplot_data <- data.metadata[!is.na(data.metadata[[metadata.name]]) & !is.na(data.metadata[[metadata.2.name]]), ]

  # Calculate median AUC per group (based only on the main grouping)
  median_values <- tapply(boxplot_data[[auc.name]], boxplot_data[[metadata.name]], median)

  # Reorder factor levels of main grouping
  boxplot_data[[metadata.name]] <- factor(boxplot_data[[metadata.name]], levels = names(sort(median_values, decreasing = FALSE)))
  
  # Ensure the fill variable is a factor
  boxplot_data[[metadata.2.name]] <- as.factor(boxplot_data[[metadata.2.name]])

  # Plot with grouping
  p <- ggplot(boxplot_data, aes(x = .data[[metadata.name]], y = .data[[auc.name]], fill = .data[[metadata.2.name]])) +
    geom_boxplot(position = position_dodge(width = 0.8), outlier.size = 0.5) +
    labs(x = metadata.name, y = auc.name, fill = metadata.2.name, title = metadata.name) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

  # Save plots
  if (is.null(filename)) filename <- paste0(output_figure_prefix, auc.name, "_AUCell_ordered_grouped_boxplot_", metadata.name, "_by_", metadata.2.name)
  ggsave(paste0(filename, ".pdf"), plot = p, width = pwidth, height = pheight, bg = "white")
  ggsave(paste0(filename, ".png"), plot = p, width = pwidth, height = pheight, dpi = 1000, bg = "white")

  return(p)
}
```

## I. Loading Seurat object

### I.1 Read previously generated Seurat object (1.6M cells)

```{r}
data.seurat <- readRDS(input_seurat_object)
data.seurat
```

### I.2 Plot

```{r fig.width=2, fig.height=3}
data.seurat$HSP90AA1 <- data.seurat@assays$RNA@data["HSP90AA1",]
plot.boxplot.aucell(data.seurat@meta.data, "HSP90AA1", "IDENT", 2, 3)
```

```{r fig.width=4, fig.height=3}
plot.boxplot.aucell.2(data.seurat@meta.data, "HSP90AA1", "IDENT", "sex", pwidth = 10)
```

```{r fig.width=4, fig.height=3}
plot.boxplot.aucell.2(data.seurat@meta.data, "HSP90AA1", "IDENT", "clinicalDiag", pwidth = 10)
```