---
title: "Emma Pan Neuro (Control + ND75KD)"
description: "Figure - Gene expression between ND75KD and Control, by annotation"
author: "Vincent Gardeux"
date: "2025/01/27"
updated: "2025/01/27"
output:
  html_document:
    df_print: paged
---

  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/gardeux/Neuro_Droso_ND75KD/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat, lib.loc = "/software/Seuratv4/lib/")) # Seurat v4.4.0, library for single-cell analysis
suppressPackageStartupMessages(library(SeuratObject, lib.loc = "/software/Seuratv4/lib/")) # For compatibility purpose
#suppressPackageStartupMessages(library(Seurat)) # Seurat v5
#suppressPackageStartupMessages(library(SeuratObject)) # Seurat v5
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(ggplot2)) # For plotting
suppressPackageStartupMessages(library(ggpubr)) # For grid plotting
suppressPackageStartupMessages(library(plotly)) # For interactive plots
suppressPackageStartupMessages(library(pheatmap)) # For nice heatmaps
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("SeuratObject"), "version", as.character(packageVersion("SeuratObject")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("ggpubr"), "version", as.character(packageVersion("ggpubr")), "\n")
cat(bold("pheatmap"), "version", as.character(packageVersion("pheatmap")), "\n")
cat(bold("plotly"), "version", as.character(packageVersion("plotly")), "\n")

# Color-blind friendly palette 3-colors
cbPalette <- c("#E69F00", "#000000", "#56B4E9")
steinPalette <- c("#80C980", "#BDADD4", "#376CB0", "#FBBF85", "#F0027E")
tissuePalette <- list(`Pan_neuro_control`="#E69F00", `Pan_neuro_ND75KD`="#56B4E9")
celltypePalette <- list(`unannotated`="#D3D3D3", `epithelial cell`="#E0AC69", `skeletal muscle of head`="#895129", `cone cell`="#FFFFB7", `outer photoreceptor cell`="#FFF060", `photoreceptor cell`="#F2DB00", `photoreceptor-like cell`="#F0C200", `adult optic chiasma glial cell`="#d896ff", `adult brain cell body glial cell`="#efbbff", `adult brain perineurial glial cell`="#BCBDDC", `adult lamina epithelial/marginal glial cell`="#9E9AC8", `adult reticular neuropil associated glial cell`="#756BB1", `optic lobe associated cortex glial cell`="#54278F", `cholinergic neuron`="#E41A1C", `gabaergic neuron`="#377EB8", `glutamatergic neuron`="#4DAF4A", `dopaminergic neuron`="#984EA3", `serotonergic neuron`="#FF7F00", `OPN neuron`="#A6CEE3", `columnar neuron T1`="#DEEBF7", `T neuron T3`="#9ECAE1", `T neuron T4/T5`="#3182BD", `distal medullary amacrine neuron Dm3`="#c6e6ba", `distal medullary amacrine neuron Dm8`="#A1D99B", `distal medullary amacrine neuron Dm9`="#31A354", `proximal medullary amacrine neuron Pm4`="#006D2C", `centrifugal neuron C2`="#7FC97F", `centrifugal neuron C3`="#BEAED4", `medullary intrinsic neuron Mi1`="#d8d5eb", `medullary intrinsic neuron Mi4`="#BCBDDC", `medullary intrinsic neuron Mi15`="#756BB1", `transmedullary neuron Tm1`="#b4dee9", `transmedullary neuron Tm2`="#bfdfe7", `transmedullary neuron Tm9`="#7eb5c4", `transmedullary neuron Tm20`="#9fcad5", `transmedullary Y neuron TmY14`="#5da1b3", `transmedullary Y neuron TmY5a`="#3c8ca2", `transmedullary Y neuron TmY8`="#1a7790", `tm5ab`="#014636", `lamina monopolar neuron L1 + L2`="#FDD3A0", `lamina monopolar neuron L3`="#FDBE85", `lamina monopolar neuron L4`="#FD8D3C", `lamina monopolar neuron L5`="#E6550D", `lamina wide-field 2 neuron`="#A63603", `lobular columnar neuron LC12`="#BC80BD", `olfactory receptor neuron`="#CCEBC5", `Poxn neuron`="#FFED6F", `hemocyte`="#999999", `kenyon cell`="#B3E2CD", `gamma kenyon cell`="#FDCDAC", `pigment cell`="#EE82EE")


# Random seed
set.seed(42)
```

## Parameters

```{r}
input_seurat_object <- "./data/Pan_neuro_integrated_FINAL.rds"
output_figure_prefix <- "./figures/Figure_2C/Boxplot_ND75KD_vs_Control_all_annotations_regulon_"
```

## I. Reading Seurat object

First step is to read the Seurat objects, previously created

```{r}
message("Loading Seurat object...")
data.seurat <- readRDS(input_seurat_object)
message(ncol(data.seurat), " cells were loaded")
message(nrow(data.seurat), " genes were loaded")
```

# Plot expression between 2 conditions

```{r, fig.width=15,fig.height=5}
# Parameters
gene <- "sima"

# Prepare
boxplot_data <- data.frame(regulon = data.seurat$SCENIC_sima_activating_regulon, annotation = data.seurat$annotation, condition = data.seurat$orig.ident)

# Plot
p <- ggplot(boxplot_data, aes(x = annotation, y = regulon, fill = condition)) +
  geom_boxplot() +
  labs(x = NULL, y = paste0(gene, " regulon activity")) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = tissuePalette)

# Save
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity.pdf"), width = 15, height = 5, bg = "white")
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity.png"), width = 15, height = 5, dpi = 1000, bg = 'white')

p
```

```{r, fig.width=15,fig.height=5}
# Parameters
gene <- "sima"

# Prepare
boxplot_data <- data.frame(regulon = data.seurat$SCENIC_sima_activating_regulon_binarized, annotation = data.seurat$annotation_clusters, condition = data.seurat$orig.ident)

# Create a table of counts
count_table <- table(boxplot_data$annotation, boxplot_data$regulon)

# Convert the table to a data frame
proportion_data <- as.data.frame(count_table)

# Rename the columns
colnames(proportion_data) <- c("annotation", "regulon", "count")

# Calculate proportions
proportion_data$proportion <- with(proportion_data, count / ave(count, annotation, FUN = sum))

# Sort annotations by the proportion of regulon == 1
sorted_annotations <- proportion_data[proportion_data$regulon == 1, ]
sorted_annotations <- sorted_annotations[order(-sorted_annotations$proportion), "annotation"]

# Convert annotation to a factor with levels sorted by the proportion of 1
proportion_data$annotation <- factor(proportion_data$annotation, levels = sorted_annotations)

# Create the stacked barplot
p <- ggplot(proportion_data, aes(x = annotation, y = proportion, fill = factor(regulon))) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("0" = "lightgrey", "1" = "darkgreen"), name = "Regulon activation") +
  labs(x = "Annotation", y = "Proportion of activated sima regulons", title = "Proportion of binarized sima regulon activity by Annotation") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Save
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity_binarized.pdf"), width = 15, height = 5, bg = "white")
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity_binarized.png"), width = 15, height = 5, dpi = 1000, bg = 'white')

p
```

```{r, fig.width=15,fig.height=5}
# Parameters
gene <- "crc"

# Prepare
boxplot_data <- data.frame(regulon = data.seurat$SCENIC_crc_activating_regulon, annotation = data.seurat$annotation, condition = data.seurat$orig.ident)

# Plot
p <- ggplot(boxplot_data, aes(x = annotation, y = regulon, fill = condition)) +
  geom_boxplot() +
  labs(x = NULL, y = paste0(gene, " regulon activity")) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = tissuePalette)

# Save
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity.pdf"), width = 15, height = 5, bg = "white")
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity.png"), width = 15, height = 5, dpi = 1000, bg = 'white')

p
```

```{r, fig.width=15,fig.height=5}
# Parameters
gene <- "crc"

# Prepare
boxplot_data <- data.frame(regulon = data.seurat$SCENIC_crc_activating_regulon_binarized, annotation = data.seurat$annotation_clusters, condition = data.seurat$orig.ident)

# Create a table of counts
count_table <- table(boxplot_data$annotation, boxplot_data$regulon)

# Convert the table to a data frame
proportion_data <- as.data.frame(count_table)

# Rename the columns
colnames(proportion_data) <- c("annotation", "regulon", "count")

# Calculate proportions
proportion_data$proportion <- with(proportion_data, count / ave(count, annotation, FUN = sum))

# Sort annotations by the proportion of regulon == 1
sorted_annotations <- proportion_data[proportion_data$regulon == 1, ]
sorted_annotations <- sorted_annotations[order(-sorted_annotations$proportion), "annotation"]

# Convert annotation to a factor with levels sorted by the proportion of 1
proportion_data$annotation <- factor(proportion_data$annotation, levels = sorted_annotations)

# Create the stacked barplot
p <- ggplot(proportion_data, aes(x = annotation, y = proportion, fill = factor(regulon))) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("0" = "lightgrey", "1" = "darkgreen"), name = "Regulon activation") +
  labs(x = "Annotation", y = "Proportion of activated crc regulons", title = "Proportion of binarized crc regulon activity by Annotation") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Save
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity_binarized.pdf"), width = 15, height = 5, bg = "white")
ggsave(p, filename = paste0(output_figure_prefix, "_boxplot_",gene,"_regulon_activity_binarized.png"), width = 15, height = 5, dpi = 1000, bg = 'white')

p
```