---
project: "Emma Pan Neuro (Control + ND75KD)"
author: "Vincent Gardeux"
date: "2025/06/06"
last_modified: "2025/06/06"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/data/gardeux/Neuro_Droso_ND75KD/")
#setwd("/data/gardeux/Neuro_Droso_ND75KD/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat)) # Seurat v5 is needed here
suppressPackageStartupMessages(library(BPCells)) # For handling big scRNAseq data in Seurat
suppressPackageStartupMessages(library(data.table)) # For writing DE gene file
suppressPackageStartupMessages(library(ggplot2)) # For plotting
suppressPackageStartupMessages(library(ggpubr)) # For grid plotting
suppressPackageStartupMessages(library(plotly)) # For interactive plots
suppressPackageStartupMessages(library(crayon)) # Just for bolding the console output :D

cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("SeuratObject"), "version", as.character(packageVersion("SeuratObject")), "\n")
cat(bold("BPCells"), "version", as.character(packageVersion("BPCells")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("ggpubr"), "version", as.character(packageVersion("ggpubr")), "\n")
cat(bold("plotly"), "version", as.character(packageVersion("plotly")), "\n")

# Random seed
set.seed(42)
```

## Parameters

*Paper:* https://www.nature.com/articles/s41586-024-07606-7

*Data repo:* Synapse

```{r}
## Parameters
input_seurat_object <- "./data/Kellis_2024/Gene Expression (snRNAseq - 10x) processed, multi-region/all_brain_regions_filt_preprocessed_scanpy_fullmatrix_excitatory.neurons.only.rds"
output_figure_prefix <- "./figures/Figure_6Q/Kellis_"
```

## Functions

```{r}
plot.boxplot.aucell.2 <- function(data.metadata, auc.name, metadata.name, metadata.2.name, pwidth = 5, pheight = 7, filename = NULL) {
  # Remove rows with NA in the relevant metadata columns
  boxplot_data <- data.metadata[!is.na(data.metadata[[metadata.name]]) & !is.na(data.metadata[[metadata.2.name]]), ]

  # Calculate median AUC per group (based only on the main grouping)
  median_values <- tapply(boxplot_data[[auc.name]], boxplot_data[[metadata.name]], median)

  # Reorder factor levels of main grouping
  boxplot_data[[metadata.name]] <- factor(boxplot_data[[metadata.name]], levels = names(sort(median_values, decreasing = FALSE)))

  # Ensure the fill variable is a factor
  boxplot_data[[metadata.2.name]] <- as.factor(boxplot_data[[metadata.2.name]])

  # Plot with grouping
  p <- ggplot(boxplot_data, aes(y = .data[[metadata.name]], x = .data[[auc.name]], fill = .data[[metadata.2.name]])) +
    geom_boxplot(position = position_dodge(width = 0.8), outlier.size = 0.5) +
    labs(y = metadata.name, x = auc.name, fill = metadata.2.name, title = metadata.name) +
    theme_minimal()

  # Save plots
  if (is.null(filename)) filename <- paste0(output_figure_prefix, auc.name, "_AUCell_ordered_grouped_boxplot_", metadata.name, "_by_", metadata.2.name)
  ggsave(paste0(filename, ".pdf"), plot = p, width = pwidth, height = pheight, bg = "white")
  ggsave(paste0(filename, ".png"), plot = p, width = pwidth, height = pheight, dpi = 1000, bg = "white")

  return(p)
}
```

## I. Loading previously generated Seurat object (436K cells)

```{r}
data.seurat <- readRDS(input_seurat_object)
data.seurat
```

## II. Plot

```{r fig.width=6, fig.height=6}
plot.boxplot.aucell.2(data.seurat@meta.data, "AUC_UPR_92", "cell_type_high_resolution", "pathAD", pwidth = 6, pheight = 6, filename = paste0(output_figure_prefix, "AUC_UPR_92_AUCell_ordered_boxplot_pathAD_by_cell_type_high_resolution_Excitatory.neurons-only"))
```
